File: app/src/main/java/com/zhhz/reader/rule/Analysis.java
package com.zhhz.reader.rule;

import static com.zhhz.reader.rule.RuleAnalysis.client;

import androidx.annotation.NonNull;

import com.alibaba.fastjson.JSONObject;
import com.zhhz.reader.bean.BookBean;
import com.zhhz.reader.util.AutoBase64;
import com.zhhz.reader.util.DiskCache;

import org.jsoup.Jsoup;
import org.mozilla.javascript.NativeArray;
import org.mozilla.javascript.NativeJavaObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Map;
import java.util.Objects;

import javax.script.ScriptException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public abstract class Analysis{

    public static String save_path = "D:\\星-阅读\\book";

    protected JSONObject json;
    //规则网站地址
    protected String url;
    //规则名字
    protected String name;
    //是否属于漫画
    protected boolean comic;
    //规则使用文本编码
    protected String charset;
    //用于标记 响应协议；
    protected String http;
    //书本信息
    protected BookBean detail;

    public Analysis(String path) throws IOException {
        this(readText(path));
    }

    public Analysis(JSONObject jsonObject) {
        this.json = jsonObject;
        this.url = jsonObject.getString("url");
        this.name = jsonObject.getString("name");
        this.comic = jsonObject.getBooleanValue("comic");
        this.charset = (String) (jsonObject.getJSONObject("search").getOrDefault("charset", "utf8"));
        if (this.charset == null) this.charset = "utf8";
        this.http = jsonObject.getJSONObject("search").getString("url").split(":")[0];
    }

    public static JSONObject readText(String path) throws IOException {
        File file = new File(path);
        if (!file.isFile()) throw new FileNotFoundException("文件未找到 -> " + file);
        FileInputStream fis;
        try {
            fis = new FileInputStream(path);
            int size = fis.available();
            byte[] bytes = new byte[size];
            if (fis.read(bytes) != size) throw new IOException("文件读取异常");
            fis.close();
            return JSONObject.parseObject(new String(bytes));
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        throw new IOException("加载异常");
    }

    public JSONObject getJson() {
        return json;
    }

    public void setJson(JSONObject json) {
        this.json = json;
    }

    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public boolean isComic() {
        return comic;
    }

    public void setComic(boolean comic) {
        this.comic = comic;
    }

    public String getCharset() {
        return charset;
    }

    public void setCharset(String charset) {
        this.charset = charset;
    }

    public String getHttp() {
        return http;
    }

    public void setHttp(String http) {
        this.http = http;
    }

    public String to_http(String relative_path, String url) {
        if (relative_path.length() > 0 && !relative_path.startsWith("http")) {
            if (relative_path.startsWith("//")) {
                return http + ":" + relative_path;
            } else if (relative_path.startsWith("/")) {
                return http + "://" + this.url + relative_path;
            } else if (!relative_path.startsWith(".")) {
                return url.substring(0, url.lastIndexOf('/')) + "/" + relative_path;
            } else {
                if (url.endsWith("/")) {
                    url = url.substring(0, url.length() - 1);
                }
                String[] arr_url = url.split("/");
                String[] arr = relative_path.split("/");
                StringBuilder sb = new StringBuilder();
                int i = 0;
                int i1 = 3;
                for (String s : arr) {
                    if (s.equals(".")) {
                        sb.append(arr_url[2]);
                    } else if (s.equals("..")) {
                        i++;
                    } else {
                        if (i < arr_url.length - 2) {
                            for (; i1 < arr_url.length - i; i1++) {
                                sb.append('/');
                                sb.append(arr_url[2 + i1]);
                            }
                        }
                    }
                }
                return sb.toString();
            }
        }


        return relative_path;
    }

    public BookBean getDetail() {
        return detail;
    }

    public void setDetail(BookBean detail) {
        this.detail = detail;
    }

    public abstract void BookSearch(String key_word, CallBack callback, String md5);

    public abstract void BookDirectory(String url, CallBack callback);

    public abstract void BookDetail(String url, CallBack callback);

    public abstract void BookChapters(BookBean book, String url, CallBack callback, Object random);

    public abstract void BookContent(String url, CallBack callback, Object random);

    public void Http(String data, CallBack callBack) {
        Http(data, callBack,false);
    }

    public void Http(String data, CallBack callBack,boolean bool) {
        if (data.contains("@post->")) {
            Http_Post(data, callBack);
        } else {
            Http_Get(data, callBack);
        }
    }

    public void Http_Post(String url, CallBack callback) {
        String header = null, data;
        MediaType mt;
        String[] arr = url.split("@post->", 2);
        if (arr[1] != null && arr[1].contains("$header")) {
            String[] ar = arr[1].split("\\$header", 2);
            data = ar[0];
            header = ar[1];
        } else {
            data = arr[1];
        }

        assert data != null;
        if (data.startsWith("{")) {
            mt = MediaType.parse("application/json");
        } else {
            mt = MediaType.parse("application/x-www-form-urlencoded;charset=" + charset);
        }

        Request.Builder builder = new Request.Builder().url(arr[0]).post(RequestBody.create(mt, data));
        if (header != null) {
            JSONObject jsonObject = JSONObject.parseObject(header);
            for (Map.Entry<String, Object> entry : jsonObject.entrySet()) {
                builder.addHeader(entry.getKey(), String.valueOf(entry.getValue()));
            }
        }
        init_header(builder);
        newCall(builder.build(), callback);
    }

    public void Http_Get(String url, CallBack callback) {
        String header = null;
        if (url.contains("$header")) {
            String[] ar = url.split("\\$header", 2);
            url = ar[0];
            header = ar[1];
        }
        Request.Builder builder = new Request.Builder().url(url);
        init_header(builder);
        if (header != null) {
            JSONObject jsonObject = JSONObject.parseObject(header);
            for (Map.Entry<String, Object> entry : jsonObject.entrySet()) {
                builder.addHeader(entry.getKey(), String.valueOf(entry.getValue()));
            }
        }
        newCall(builder.build(), callback);
    }

    private void init_header(Request.Builder builder) {
        if (json.get("header") != null) {
            String h = this.json.getString("header");
            if (h.indexOf("js@") == 0) {
                try {
                    h = JsToJava(DiskCache.SCRIPT_ENGINE.eval(AutoBase64.decodeToString(h.substring(3))));
                } catch (ScriptException e) {
                    h = "{}";
                }
            }
            JSONObject header = JSONObject.parseObject(h);
            for (Map.Entry<String, Object> entry : header.entrySet()) {
                builder.addHeader(entry.getKey(), String.valueOf(entry.getValue()));
            }
            header.clear();
        }
    }

    private void newCall(Request request, CallBack callback) {
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                callback.run(null, e.getMessage(), null);
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                if (response.code() == 200 && response.body() != null) {
                    MediaType contentType = response.body().contentType();
                    String charset = null;
                    if (contentType != null && contentType.charset() != null) {
                        charset = Objects.requireNonNull(contentType.charset()).name();
                    }
                    if (charset == null) {
                        charset = Analysis.this.charset;
                    }
                    String s = new String(Objects.requireNonNull(response.body()).bytes(), charset);
                    DiskCache.FileSave(DiskCache.path, call, s);
                    callback.run(Jsoup.parse(s), null, null);
                } else {
                    callback.run(null, response.message(), null);
                }
            }
        });
    }

    public String JsToJava(Object value) {
        if (value == null) return "";
        String str;
        if (value instanceof NativeArray) {
            NativeArray array = (NativeArray) value;
            StringBuilder stringBuilder = new StringBuilder();
            for (Object o : array) {
                stringBuilder.append(o).append("\n");
            }
            stringBuilder.delete(stringBuilder.length() - 1, stringBuilder.length());
            str = stringBuilder.toString();
        } else if (value.getClass() == NativeJavaObject.class) {
            str = ((NativeJavaObject) value).unwrap().toString();
        } else {
            str = value.toString();
        }
        return str;
    }

    @FunctionalInterface
    public interface CallBack {
        /**
         * @param data  数据
         * @param msg   报错提示
         * @param label 随机数，用于标记
         */
        void run(Object data, Object msg, Object label);
    }
}

File: app/src/main/java/com/zhhz/reader/rule/Analysis.kt
package com.zhhz.reader.rule

import cn.hutool.core.codec.Base64
import cn.hutool.core.util.ObjectUtil
import com.alibaba.fastjson.JSON
import com.alibaba.fastjson.JSONObject
import com.zhhz.reader.bean.BookBean
import com.zhhz.reader.bean.HttpResponseBean
import com.zhhz.reader.bean.rule.RuleJsonBean
import com.zhhz.reader.util.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.*
import java.net.InetAddress
import java.net.MalformedURLException
import java.net.URL
import java.nio.charset.Charset
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit
import javax.script.ScriptException
import javax.script.SimpleBindings

abstract class Analysis(var json: RuleJsonBean): JsExtensionClass {

    //规则网站地址
    var url: String = json.url

    //规则名字
    var name: String = json.name

    //是否属于漫画
    var isComic: Boolean = json.comic

    //规则使用文本编码
    var charset: String

    //用于标记 响应协议；
    private var http: String

    //书本信息
    var detail: BookBean? = null

    //用于存储数据
    private val shareMap = HashMap<Any, Any>()
    private var client: OkHttpClient? = null

    fun ajax(
        url: String,
        method: String,
        mediaType: String,
        header: String,
        data: String
    ): String? {
        return (if (client == null) share_client else client)?.let { super.ajax(it, url, method, mediaType, header, data) }
    }

    fun ajax(
        url: String,
        method: String,
        mediaType: String
    ): String? {
        return (if (client == null) share_client else client)?.let { ajax(it, url, method, mediaType) }
    }

    constructor(path: String) : this(readText(path))

    init {
        var share = true
        if (isHaveSearch) {
            charset = json.search.charset ?: json.charset
            http = json.search.url.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
        } else {
            http = "http"
            charset = json.charset
        }
        if (charset.isEmpty()) charset = "utf8"
        val builder = OkHttpClient.Builder().connectTimeout(30,TimeUnit.SECONDS).readTimeout(30,TimeUnit.SECONDS).writeTimeout(30,TimeUnit.SECONDS)
        if (json.cache) {
            share = false
            builder.addInterceptor(DiskCache.interceptor)
        }
        if (json.cookieJar) {
            share = false
            builder.cookieJar(CookiesManager())
        }
        if (json.dns.isNotEmpty()) {
            share = false
            builder.dns(object :Dns{
                override fun lookup(hostname: String): List<InetAddress> {
                    return JSON.parseArray(json.dns).map {
                        InetAddress.getByName(it as String)
                    }
                }
            })
        }

        if (ObjectUtil.isEmpty(json.init)) {
            client = if (share) {
                share_client
            } else {
                builder.build()
            }
        } else {
            val bindings = SimpleBindings()
            bindings["builder"] = builder
            bindings["xlua_rule"] = this
            bindings["java"] = this
            client = try {
                val obj = DiskCache.SCRIPT_ENGINE.eval(Base64.decodeStr(json.init), bindings)
                if (obj is OkHttpClient.Builder) {
                    obj.build()
                } else {
                    builder.build()
                }
            } catch (e: ScriptException) {
                log(e)
                builder.build()
            }
        }
    }

    fun log(e: Exception) {
        val byteArrayOutputStream = ByteArrayOutputStream()
        e.printStackTrace(PrintStream(byteArrayOutputStream))
        log(byteArrayOutputStream.toString())
    }

    fun log(s: Any) {
        if (logError != null) {
            logError!!.log(s.toString())
        }
    }

    //设置共享数据
    fun setShareValue(key: Any, value: Any) {
        shareMap[key] = value
    }

    //获取共享数据
    fun getShareValue(key: Any): Any {
        return shareMap[key]?:""
    }

    fun toAbsoluteUrl(relativePath: String, absoluteBasePath: String = url): String {
        if (relativePath.equals("skip", ignoreCase = true)) return relativePath
        return if (relativePath.isNotEmpty() && !relativePath.startsWith("http")) {
            try {

                val absoluteUrl = if (!absoluteBasePath.startsWith("http")){
                    URL("$http://$url")
                } else {
                    URL(absoluteBasePath)
                }

                URL(absoluteUrl, relativePath).toString()
            } catch (e: MalformedURLException) {
                e.printStackTrace()
                ""
            }
        } else relativePath
    }

    /**
     * 判断是否有搜索
     * @return boolean
     */
    val isHaveSearch: Boolean
        get() = ObjectUtil.isNotNull(json.search) && ObjectUtil.isNotEmpty(json.search.url)

    abstract fun bookSearch(keyWord: String,page: Int, callback: AnalysisCallBack.SearchCallBack, label: String)
    abstract fun bookLeaderboard(leaderboardUrl: String, page: Int, callback: AnalysisCallBack.LeaderboardCallBack, label: String)
    abstract fun bookDetail(url: String, callback: AnalysisCallBack.DetailCallBack)
    abstract fun bookDirectory(url: String, callback: AnalysisCallBack.DirectoryCallBack)
    fun bookChapters(book: BookBean, url: String, callback: AnalysisCallBack.ContentCallBack, label: Any) {
        //val file = File(DiskCache.path + File.separator + "book" + File.separator + book.book_id + File.separator + "book_chapter" + File.separator + url.substring(url.lastIndexOf('/') + 1))
        val file = File(DiskCache.path + File.separator + "book" + File.separator + book.bookId + File.separator + "book_chapter" + File.separator + StringUtil.getMD5(url))

        if (!Objects.requireNonNull(file.parentFile).isDirectory) {
            if (!file.parentFile?.mkdirs()!!) {
                callback.accept(null, label)
            }
        }
        if (file.isFile) {
            try {
                FileInputStream(file).use { fis ->
                    val size = fis.available()
                    val bytes = ByteArray(size)
                    if (fis.read(bytes) == size) {
                        val h = HttpResponseBean()
                        h.isStatus = true
                        h.data = String(bytes)
                        callback.accept(h, label)
                        return
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        bookContent(url, AnalysisCallBack.ContentCallBack { data: HttpResponseBean, tag: Any ->
            if (!data.isStatus) {
                callback.accept(data, tag)
                return@ContentCallBack
            }
            try {
                FileOutputStream(file).use { fos -> fos.write(data.data.toByteArray()) }
            } catch (e: IOException) {
                e.printStackTrace()
            }
            callback.accept(data, tag)
        }, label)
    }

    abstract fun bookContent(url: String, callback: AnalysisCallBack.ContentCallBack, label: Any)

    fun http(url: String, callBack: AnalysisCallBack.CallBack) {
        Coroutine.async {
        //url等于skip跳过请求
        if (url.equals("skip", ignoreCase = true)) {
            val httpResponseBean = HttpResponseBean()
            httpResponseBean.isStatus = true
            callBack.accept(httpResponseBean)
            return@async
        }
        if (url.contains("@post->")) {
            httpPost(url, callBack)
        } else {
            httpGet(url, callBack)
        }
        }.onError {
            it.printStackTrace()
            val httpResponseBean = HttpResponseBean()
            httpResponseBean.isStatus = false
            callBack.accept(httpResponseBean)
        }
    }

    fun httpPost(url: String, callback: AnalysisCallBack.CallBack) {
        var header: String? = null
        val data: String
        val arr= url.split("@post->".toRegex(), limit = 2)
        if (arr[1].isNotEmpty() && arr[1].contains("\$header")) {
            val ar = arr[1].split("\\\$header".toRegex(), limit = 2)
            data = ar[0]
            header = ar[1]
        } else {
            data = arr[1]
        }
        val mt: MediaType? = if (data.startsWith("{")) {
            "application/json; charset=$charset".toMediaTypeOrNull()
        } else {
            "application/x-www-form-urlencoded;charset=$charset".toMediaTypeOrNull()
        }

        val builder: Request.Builder = Request.Builder().url(arr[0]).post(data.toRequestBody(mt))
        if (header != null) {
            val jsonObject = JSONObject.parseObject(header)
            for ((key, value) in jsonObject) {
                builder.addHeader(key, value.toString())
            }
        }
        initHeader(builder,arr[0])
        newCall(builder.build(), callback)
    }

    /**
     * GET请求
     *
     * @param url      地址
     * @param callback 回调事件
     */
    fun httpGet(url: String, callback: AnalysisCallBack.CallBack) {
        var header: String? = null
        val urlTemp = if (url.contains("\$header")) {
            val ar = url.split("\\\$header".toRegex(), limit = 2)
            header = ar[1]
            ar[0]
        } else {
            url
        }
        val builder: Request.Builder = Request.Builder().url(urlTemp)
        initHeader(builder,url)
        if (header != null) {
            val jsonObject = JSONObject.parseObject(header)
            for ((key, value) in jsonObject) {
                builder.addHeader(key, value.toString())
            }
        }
        newCall(builder.build(), callback)
    }

    private fun initHeader(builder: Request.Builder,httpUrl: String) {
        if (json.header != null) {
            var h = json.header
            if (h.contains("^js@|@js:".toRegex())) {
                try {
                    val bindings = SimpleBindings()
                    bindings["xlua_rule"] = this
                    bindings["java"] = this
                    bindings["url"] = httpUrl

                    h = jsToJavaObject(
                        DiskCache.SCRIPT_ENGINE.eval(
                            Base64.decodeStr(h.split("js@|@js:".toRegex())[1]),
                            bindings
                        )
                    )
                } catch (e: ScriptException) {
                    e.printStackTrace()
                    h = "{}"
                }
            }
            if (h.isNotEmpty()) {
                val header = JSONObject.parseObject(h)
                for ((key, value) in header) {
                    builder.addHeader(key, value.toString())
                }
                header.clear()
            }
        }
    }

    private fun newCall(request: Request, callback: AnalysisCallBack.CallBack) {
        val httpResponseBean = HttpResponseBean()
        httpResponseBean.requestHeader = request.headers
        (client ?: share_client).newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                httpResponseBean.isStatus = false
                httpResponseBean.error = e.message.toString()
                callback.accept(httpResponseBean)
            }
            override fun onResponse(call: Call, response: Response) {
                httpResponseBean.isStatus = response.code == 200
                var charset: String? = null
                if (ObjectUtil.isNotNull(response.body)) {
                    val contentType = response.body?.contentType()
                    if (contentType?.charset() != null) {
                        charset = contentType.charset()?.name()
                    }
                }

                var s = ""
                try {
                    s = if (this@Analysis.charset == charset) {
                        response.body?.string()
                    } else {
                        charset = this@Analysis.charset
                        response.body?.bytes()?.toString(Charset.forName(charset))
                    }.orEmpty()
                    if (httpResponseBean.isStatus) DiskCache.FileSave(DiskCache.path, call, s.toByteArray(charset = Charset.forName(charset)))
                } catch (e: IOException) {
                    httpResponseBean.isStatus = false
                    httpResponseBean.error = e.message!!
                    log(e)
                    e.printStackTrace()
                }
                httpResponseBean.data = s
                callback.accept(httpResponseBean)
                response.close()
            }
        })
    }


    companion object {
        var share_client: OkHttpClient = OkHttpClient.Builder().addInterceptor(DiskCache.interceptor).connectTimeout(30,TimeUnit.SECONDS).readTimeout(30,TimeUnit.SECONDS).writeTimeout(30,TimeUnit.SECONDS).build()
        private var logError: AnalysisCallBack.LogError? = null
        @JvmStatic
        fun setLogError(error: AnalysisCallBack.LogError) {
            logError = error
        }

        @Throws(IOException::class)
        fun readText(path: String): RuleJsonBean {
            return CompletableFuture.supplyAsync {
                try {
                    return@supplyAsync JSON.parseObject(FileUtil.readFile(path),RuleJsonBean::class.java)
                } catch (e: IOException) {
                    LogUtil.error(e)
                    throw FileNotFoundException("文件读取异常 -> $path")
                }
            }.join()
        }
    }
}